### DISCUSSION TITLE
Specifying the object address scheme for devices?

### DISCUSSION THREAD

**Discussion by ncryer** [stepfunc/dnp3#305] [Q&A] [answered]
Hello everyone, thanks for making this library first and foremost.  
 
I'm tasked with creating some monitoring software for DNP3 systems, to which end I'm creating a simulator to iterate against. I've done this for the IEC104 protocol before, and in that case there was a straightforward way to define the object address layout for the system in question - in other words, all the particular sensors/devices registered to the system - but I can't seem to find anything similar in this library. 
 
Is there a way to do this that I have missed? 
 
Thanks a lot

**Comment by ncryer** [comment]
Well, answering my own question by doing a bit of RTFM, it's specified here for outstations: https://docs.stepfunc.io/dnp3/1.4.1/guide/docs/api/outstation/database 
 
I suppose my next questions are: 
 
- Are Master stations considered stateless in this protocol? In other words, are any and all devices situated in an outstation configuration? 
- For my application I'm (supposed to be) reading off alarm codes from the Master station. Is it correct to assume that this can be achieved by creating a `MasterChannel` (in the client) with a TCP connection to whatever endpoint the actual to-be-monitored Master Station exposes, and then using that connection to filter messages? 

Hope noob questions are alright here

**Comment by jadamcrain** [comment, marked as answer]
`Are Master stations considered stateless in this protocol? In other words, are any and all devices situated in an outstation configuration?`

Pretty much nothing is stateless in DNP3.  The outstation is worse than the master in this regard though. I'm not sure what you mean by the devices situated in an outstation configuration. The outstation (server) is the remote device that the master (client) connects to, reads values from, and writes controls to.

`..and then using that connection to filter messages?`

Yes, on top of the `MasterChannel` there's a concept of an [Association](https://docs.rs/dnp3/1.4.1/dnp3/master/struct.AssociationHandle.html).  A channel can talk to N devices connected to that channel. Over a serial port, this can be physically separate devices on the same RS-485 bus. The communication is usually just 1:1 on TCP.

[ReadHandler::handle_binary_input](https://docs.rs/dnp3/1.4.1/dnp3/master/trait.ReadHandler.html#method.handle_binary_input) will be called when measurement data is received from the outstation in question.

**Reply by ncryer** [reply]
Thanks a lot for taking the time! 

The reason I ask about "outstation configuration" is because it looks like I can only get a `Database` handle either in a `ControlHandler` or with an `Outstation` class. So my mental model (which is probably wrong?) is that only outstations can have state, and state is relayed to the Master or updated due to writes from the Master, but the Master itself doesn't hold any.

**Reply by jadamcrain** [reply]
I see what you mean by "state" now.  Your mental model is correct. The outstation holds the measurement values in internal data-structures because it needs them in a certain format to satisfy requests from the master.

The master just receives the values and immediately passes them to the user application.  Most real applications will do something besides printing the values, e.g. transform them in some way (index -> human readable tag name) and then write them to a database. But yes, the master side of the library doesn't hold on to them in any way.

Are you building something for a product or is this an academic or prototyping exercise?

**Reply by ncryer** [reply]
Thanks so much for the answer, that's super helpful! 

> Are you building something for a product or is this an academic or prototyping exercise? 
 
This is a prototyping exercise as the first step towards a product for one of our clients. It's not sold yet so I can't give any guarantees, but let me know how I can repay your kindness (with money) and I'll be happy to make the arrangements if the client goes for it.  
 
Edit: I just read LICENSE.txt, it seems I would need a commercial license if the time comes.

**Reply by jadamcrain** [reply]
Hi @ncryer. That's right. If you're interested knowing the pricing/etc just email info@stepfunc.io and I can send you this information.
